# 18. 예외처리
---
* 프로그램이 동작하는 과정 중 잘못된 입력 값이 전달되거나
* 읽어야 하는 파일이 존재하지 않거나
* 배열의 인덱스를 잘못 전달했거나
* Null 인스턴스에 접근을 하려했거나 등등의 이유로 오류가 발생하는데
* 이것을 "예외"라고 한다.
* 예외가 발생한 경우 Java 프로그램은 JVM에 의해서 강제 종료된다.

* 예외 처리는 예외가 발생하더라도 프로그램이 종료되지 않도록 하는 방법이면서
* 예외가 발생하지 않도록 하는 방어코딩의 방법이다.

* Java에서 가장 많이 발생하는 예외 목록
  1. NullPointerException
      * Null Instance에 접근하려 함
  2. IndexOutOfBoundsException
      * List의 인덱스 범위를 벗어남
  3. ArrayIndexOutOfBoundsException
      * 배열의 인덱스 범위를 벗어남
  4. NumberFormatException
      * 숫자가 아닌 문자를 숫자로 변환하려 함

## if를 활용한 예외처리
* 예외를 처리하는 가장 좋은 방법은 예외를 발생시키지 않는 것이다.
* 아래 예외 목록은 if를 활용해 예외를 방지할 수 있다.
  
  1. NullPointerException
       * Null Instance에 접근하려 함
       * Instance가 Null인지 먼저 확인 함
  2. IndexOutOfBoundsException
       * List의 인덱스 범위를 벗어남
       * 찾으려는 인덱스가 존재하는지 먼저 확인 함
  3. ArrayIndexOutOfBoundsException
       * 배열의 인덱스 범위를 벗어남
       * 찾으려는 인덱스가 존재하는지 먼저 확인 함
  4. NumberFormatException
       * 숫자가 아닌 문자를 숫자로 변환하려 함
       * 문자가 숫자형태인지 확인 함

## try - catch - finally 예외처리
* if로 예외 처리할 수 없거나 애매할 경우 try~catch~finally를 사용한다.
```java
try {
    // ▼ NumberFormatException 발생
    int number = Integer.parseInt("ABC"); 
    System.out.println(number); // 이 코드는 실행되지 않는다.
}
// ▼ NumberFormatException이 발생했을 때 catch가 실행된다.
catch ( NumberFormatException e ) {
    // try에서 예외가 발생하면 catch가 실행된다.
    System.out.println("에러가 발생했습니다. " + e.getMessage());
    e.printStackTrace(); // 에러내용 모두 출력
}
finally {
    // 예외 발생 여부와 관계없이 무조건 실행된다.
    System.out.println("처리가 완료됐습니다.");
}
```
* Java의 예외 종류는 크게 두 종류가 있다.
<img width="750" alt="image" src="https://github.com/user-attachments/assets/9af71a9c-f448-40c3-8903-3a66fa17e69e">
* if로 처리할 수 없는 예외들은 try ~ catch ~ finally를 사용한다.

## 사용자 예외 발생시키기
* Java에서 제공하는 내장 Exception은 Java Application이 동작할 때,
  코드의 조건을 만족하지 못했을 때 지정된 Exception이 발생하도록 되어 있다.
* 사용자 예외는 항상 RuntimeException을 상속받아 생성한다.
  
```java
public class InvalidUserIdException extends RuntimeException {
}
```
* 생성자가 존재하는 클래스를 상속 받으면 Sub Class에서 Super Class의 생성자를 호출해주어야 한다.
 
